kv/kvserver: TestWedgedReplicaDetection failed

kv/kvserver.TestWedgedReplicaDetection [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5313837&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5313837&tab=artifacts#/) on release-22.1 @ [34e6fcfdc8f3155831305ab4a78f960aaad3e7bc](https://github.com/cockroachdb/cockroach/commits/34e6fcfdc8f3155831305ab4a78f960aaad3e7bc):


```
=== RUN   TestWedgedReplicaDetection
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestWedgedReplicaDetection38094081
    test_log_scope.go:80: use -show-logs to present logs inline
    client_raft_test.go:2391: expected follower to still be considered active
    panic.go:642: -- test log scope end --
--- FAIL: TestWedgedReplicaDetection (3.07s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWedgedReplicaDetection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-16192

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -2648,6 +2648,7 @@ func TestWedgedReplicaDetection(t *testing.T) {
 	const numReplicas = 3
 
 	ctx := context.Background()
+	manual := hlc.NewHybridManualClock()
 	tc := testcluster.StartTestCluster(t, numReplicas,
 		base.TestClusterArgs{
 			ReplicationMode: base.ReplicationManual,
@@ -2657,15 +2658,30 @@ func TestWedgedReplicaDetection(t *testing.T) {
 					// this test doesn't expect.
 					RaftElectionTimeoutTicks: 100000,
 				},
+				Knobs: base.TestingKnobs{
+					Server: &server.TestingKnobs{
+						WallClock: manual,
+					},
+				},
 			},
 		})
 	defer tc.Stopper().Stop(ctx)
+
+	// Pause the manual clock so that we can carefully control the perceived
+	// timing of the follower replica's activity.
+	manual.Pause()
 
 	key := []byte("a")
 	tc.SplitRangeOrFatal(t, key)
 	tc.AddVotersOrFatal(t, key, tc.Targets(1, 2)...)
 
+	// Do a write; we'll use it to determine when the dust has settled.
+	_, err := tc.Servers[0].DB().Inc(ctx, key, 1)
+	require.Nil(t, err)
+	tc.WaitForValues(t, key, []int64{1, 1, 1})
+
 	leaderRepl := tc.GetRaftLeader(t, key)
+	leaderClock := leaderRepl.Clock()
 	followerRepl := func() *kvserver.Replica {
 		for i := range tc.Servers {
 			repl := tc.GetFirstStoreFromServer(t, i).LookupReplica(key)
@@ -2693,18 +2709,20 @@ func TestWedgedReplicaDetection(t *testing.T) {
 	wg.Wait()
 	defer followerRepl.RaftUnlock()
 
-	// TODO(andrei): The test becomes flaky with a lower threshold because the
-	// follower is considered inactive just below. Figure out how to switch the
-	// test to a manual clock. The activity tracking for followers uses the
-	// physical clock.
+	// inactivityThreshold is the test's duration of inactivity after which the
+	// follower replica is considered inactive. In practice, this is set to the
+	// range lease duration.
 	inactivityThreshold := time.Second
+
+	// Increment the clock to be close to inactivityThreshold, but not past it.
+	manual.Increment(inactivityThreshold.Nanoseconds() - 1)
 
 	// Send a request to the leader replica. followerRepl is locked so it will
 	// not respond.
 	value := []byte("value")
 	ba := &kvpb.BatchRequest{}
 	ba.Add(putArgs(key, value))
-	if err := ba.SetActiveTimestamp(tc.Servers[0].Clock()); err != nil {
+	if err := ba.SetActiveTimestamp(leaderClock); err != nil {
 		t.Fatal(err)
 	}
 	if _, pErr := leaderRepl.Send(ctx, ba); pErr != nil {
@@ -2713,7 +2731,7 @@ func TestWedgedReplicaDetection(t *testing.T) {
 
 	// The follower should still be active.
 	followerID := followerRepl.ReplicaID()
-	if !leaderRepl.IsFollowerActiveSince(ctx, followerID, inactivityThreshold) {
+	if !leaderRepl.IsFollowerActiveSince(followerID, leaderClock.PhysicalTime(), inactivityThreshold) {
 		t.Fatalf("expected follower to still be considered active")
 	}
 
@@ -2722,6 +2740,9 @@ func TestWedgedReplicaDetection(t *testing.T) {
 	// would bump the last active timestamp on the leader. Because of this,
 	// we check whether the follower is eventually considered inactive.
 	testutils.SucceedsSoon(t, func() error {
+		// Increment the clock to past inactivityThreshold.
+		manual.Increment(inactivityThreshold.Nanoseconds() + 1)
+
 		// Send another request to the leader replica. followerRepl is locked
 		// so it will not respond.
 		if _, pErr := leaderRepl.Send(ctx, ba); pErr != nil {
@@ -2729,7 +2750,7 @@ func TestWedgedReplicaDetection(t *testing.T) {
 		}
 
 		// The follower should no longer be considered active.
-		if leaderRepl.IsFollowerActiveSince(ctx, followerID, inactivityThreshold) {
+		if leaderRepl.IsFollowerActiveSince(followerID, leaderClock.PhysicalTime(), inactivityThreshold) {
 			return errors.New("expected follower to be considered inactive")
 		}
 		return nil

--- a/pkg/kv/kvserver/helpers_test.go
+++ b/pkg/kv/kvserver/helpers_test.go
@@ -48,7 +48,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/metric"
 	"github.com/cockroachdb/cockroach/pkg/util/quotapool"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
-	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 	"go.etcd.io/raft/v3"
@@ -416,11 +415,11 @@ func (r *Replica) NumPendingProposals() int {
 }
 
 func (r *Replica) IsFollowerActiveSince(
-	ctx context.Context, followerID roachpb.ReplicaID, threshold time.Duration,
+	followerID roachpb.ReplicaID, now time.Time, threshold time.Duration,
 ) bool {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
-	return r.mu.lastUpdateTimes.isFollowerActiveSince(followerID, timeutil.Now(), threshold)
+	return r.mu.lastUpdateTimes.isFollowerActiveSince(followerID, now, threshold)
 }
 
 // GetTSCacheHighWater returns the high water mark of the replica's timestamp

--- a/pkg/kv/kvserver/replica_proposal_quota.go
+++ b/pkg/kv/kvserver/replica_proposal_quota.go
@@ -21,7 +21,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/quotapool"
-	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"go.etcd.io/raft/v3"
 	"go.etcd.io/raft/v3/tracker"
@@ -89,6 +88,7 @@ var logSlowRaftProposalQuotaAcquisition = quotapool.OnSlowAcquisition(
 func (r *Replica) updateProposalQuotaRaftMuLocked(
 	ctx context.Context, lastLeaderID roachpb.ReplicaID,
 ) {
+	now := r.Clock().PhysicalTime()
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
@@ -119,7 +119,7 @@ func (r *Replica) updateProposalQuotaRaftMuLocked(
 				logSlowRaftProposalQuotaAcquisition,
 			)
 			r.mu.lastUpdateTimes = make(map[roachpb.ReplicaID]time.Time)
-			r.mu.lastUpdateTimes.updateOnBecomeLeader(r.mu.state.Desc.Replicas().Descriptors(), timeutil.Now())
+			r.mu.lastUpdateTimes.updateOnBecomeLeader(r.mu.state.Desc.Replicas().Descriptors(), now)
 			r.mu.replicaFlowControlIntegration.onBecameLeader(ctx)
 			r.mu.lastProposalAtTicks = r.mu.ticks // delay imminent quiescence
 		} else if r.mu.proposalQuota != nil {
@@ -143,9 +143,8 @@ func (r *Replica) updateProposalQuotaRaftMuLocked(
 	}
 
 	// We're still the leader.
-
 	// Find the minimum index that active followers have acknowledged.
-	now := timeutil.Now()
+
 	// commitIndex is used to determine whether a newly added replica has fully
 	// caught up.
 	commitIndex := kvpb.RaftIndex(status.Commit)

--- a/pkg/kv/kvserver/replica_raft.go
+++ b/pkg/kv/kvserver/replica_raft.go
@@ -601,7 +601,7 @@ func (r *Replica) stepRaftGroup(req *kvserverpb.RaftMessageRequest) error {
 			wakeLeader := hasLeader && !fromLeader
 			r.maybeUnquiesceLocked(wakeLeader, false /* mayCampaign */)
 		}
-		r.mu.lastUpdateTimes.update(req.FromReplica.ReplicaID, timeutil.Now())
+		r.mu.lastUpdateTimes.update(req.FromReplica.ReplicaID, r.Clock().PhysicalTime())
 		switch req.Message.Type {
 		case raftpb.MsgPreVote, raftpb.MsgVote:
 			// If we receive a (pre)vote request, and we find our leader to be dead or
@@ -1313,7 +1313,7 @@ func (r *Replica) tick(
 	// This is likely unintentional, and the leader should likely consider itself
 	// live even when quiesced.
 	if r.isRaftLeaderRLocked() {
-		r.mu.lastUpdateTimes.update(r.replicaID, timeutil.Now())
+		r.mu.lastUpdateTimes.update(r.replicaID, r.Clock().PhysicalTime())
 	}
 
 	r.mu.ticks++

--- a/pkg/kv/kvserver/replica_raft_quiesce.go
+++ b/pkg/kv/kvserver/replica_raft_quiesce.go
@@ -22,7 +22,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/util/envutil"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
-	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"go.etcd.io/raft/v3"
 	"go.etcd.io/raft/v3/raftpb"
 	"go.etcd.io/raft/v3/tracker"
@@ -94,7 +93,7 @@ func (r *Replica) maybeUnquiesceLocked(wakeLeader, mayCampaign bool) bool {
 	st := r.raftSparseStatusRLocked()
 	if st.RaftState == raft.StateLeader {
 		r.mu.lastUpdateTimes.updateOnUnquiesce(
-			r.mu.state.Desc.Replicas().Descriptors(), st.Progress, timeutil.Now())
+			r.mu.state.Desc.Replicas().Descriptors(), st.Progress, r.Clock().PhysicalTime())
 
 	} else if st.RaftState == raft.StateFollower && st.Lead != raft.None && wakeLeader {
 		// Propose an empty command which will wake the leader.

--- a/pkg/util/hlc/hlc.go
+++ b/pkg/util/hlc/hlc.go
@@ -437,15 +437,7 @@ func (c *Clock) PhysicalNow() int64 {
 }
 
 // PhysicalTime returns a time.Time struct using the local wall time.
-func (c *Clock) PhysicalTime() time.Time {
-	// NOTE: We don't go through c.physicalNanos() if c.timeSource is set in order
-	// to preserve the monotonic clock reading that the timeSource might provide
-	// inside its instants.
-	if c.wallClock != nil {
-		return c.wallClock.Now()
-	}
-	return timeutil.Unix(0, c.wallClock.Now().UnixNano())
-}
+func (c *Clock) PhysicalTime() time.Time { return c.wallClock.Now() }
 
 // Update takes a hybrid timestamp, usually originating from an event
 // received from another member of a distributed system. The clock is

